File: tools/architecture_guardrails.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any


DEFAULT_CONFIG_PATH = Path(__file__).with_name("architecture_guardrails.json")


class GuardrailConfigError(RuntimeError):
    pass


@dataclass(frozen=True)
class LineLimitRule:
    path: str
    max_lines: int
    owner: str
    review_note: str


@dataclass(frozen=True)
class ForbiddenPathRule:
    path: str
    owner: str
    reason: str


@dataclass(frozen=True)
class GuardrailConfig:
    line_limits: tuple[LineLimitRule, ...]
    forbidden_paths: tuple[ForbiddenPathRule, ...]


@dataclass(frozen=True)
class Violation:
    code: str
    path: str
    message: str


def require_non_empty_text(value: Any, *, field_name: str, context: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise GuardrailConfigError(f"{context} is missing required text field {field_name}")
    return value.strip()


def require_positive_int(value: Any, *, field_name: str, context: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value <= 0:
        raise GuardrailConfigError(f"{context} has invalid positive integer field {field_name}")
    return value


def parse_line_limit_rule(raw: Any, index: int) -> LineLimitRule:
    context = f"line_limits[{index}]"
    if not isinstance(raw, dict):
        raise GuardrailConfigError(f"{context} must be an object")
    return LineLimitRule(
        path=require_non_empty_text(raw.get("path"), field_name="path", context=context),
        max_lines=require_positive_int(raw.get("max_lines"), field_name="max_lines", context=context),
        owner=require_non_empty_text(raw.get("owner"), field_name="owner", context=context),
        review_note=require_non_empty_text(raw.get("review_note"), field_name="review_note", context=context),
    )


def parse_forbidden_path_rule(raw: Any, index: int) -> ForbiddenPathRule:
    context = f"forbidden_paths[{index}]"
    if not isinstance(raw, dict):
        raise GuardrailConfigError(f"{context} must be an object")
    return ForbiddenPathRule(
        path=require_non_empty_text(raw.get("path"), field_name="path", context=context),
        owner=require_non_empty_text(raw.get("owner"), field_name="owner", context=context),
        reason=require_non_empty_text(raw.get("reason"), field_name="reason", context=context),
    )


def load_config(config_path: Path) -> GuardrailConfig:
    try:
        raw = json.loads(config_path.read_text(encoding="utf-8"))
    except FileNotFoundError as err:
        raise GuardrailConfigError(f"guardrail config not found: {config_path}") from err
    except json.JSONDecodeError as err:
        raise GuardrailConfigError(f"guardrail config is not valid json: {config_path}: {err}") from err

    if not isinstance(raw, dict):
        raise GuardrailConfigError("guardrail config root must be an object")

    raw_line_limits = raw.get("line_limits", [])
    raw_forbidden_paths = raw.get("forbidden_paths", [])
    if not isinstance(raw_line_limits, list):
        raise GuardrailConfigError("line_limits must be an array")
    if not isinstance(raw_forbidden_paths, list):
        raise GuardrailConfigError("forbidden_paths must be an array")

    return GuardrailConfig(
        line_limits=tuple(parse_line_limit_rule(item, index) for index, item in enumerate(raw_line_limits)),
        forbidden_paths=tuple(parse_forbidden_path_rule(item, index) for index, item in enumerate(raw_forbidden_paths)),
    )


def count_lines(path: Path) -> int:
    return len(path.read_text(encoding="utf-8", errors="replace").splitlines())


def evaluate_guardrails(repo_root: Path, config: GuardrailConfig) -> list[Violation]:
    violations: list[Violation] = []

    for rule in config.line_limits:
        candidate = repo_root / rule.path
        if not candidate.is_file():
            violations.append(
                Violation(
                    code="GUARDED_FILE_MISSING",
                    path=rule.path,
                    message=(
                        f"guarded file is missing; either update the architecture guardrail or restore the file "
                        f"(owner: {rule.owner}; review note: {rule.review_note})"
                    ),
                )
            )
            continue

        line_count = count_lines(candidate)
        if line_count > rule.max_lines:
            violations.append(
                Violation(
                    code="MAX_LINES_EXCEEDED",
                    path=rule.path,
                    message=(
                        f"{line_count} lines exceeds max {rule.max_lines} "
                        f"(owner: {rule.owner}; review note: {rule.review_note})"
                    ),
                )
            )

    for rule in config.forbidden_paths:
        candidate = repo_root / rule.path
        if candidate.exists():
            violations.append(
                Violation(
                    code="FORBIDDEN_PATH_PRESENT",
                    path=rule.path,
                    message=f"forbidden path is present (owner: {rule.owner}; reason: {rule.reason})",
                )
            )

    return violations


def format_violations(violations: list[Violation]) -> str:
    lines = ["Architecture guardrails failed:"]
    for violation in violations:
        lines.append(f"- [{violation.code}] {violation.path}: {violation.message}")
    return "\n".join(lines)


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Fail when protected files regrow or forbidden paths reappear.")
    parser.add_argument(
        "--repo-root",
        type=Path,
        default=Path(__file__).resolve().parent.parent,
        help="Path to the repository root. Defaults to the parent of tools/.",
    )
    parser.add_argument(
        "--config",
        type=Path,
        default=DEFAULT_CONFIG_PATH,
        help="Path to the guardrail config json file.",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        config = load_config(args.config)
    except GuardrailConfigError as err:
        print(f"Architecture guardrails config error: {err}", file=sys.stderr)
        return 2

    violations = evaluate_guardrails(args.repo_root.resolve(), config)
    if violations:
        print(format_violations(violations), file=sys.stderr)
        return 1

    print(
        "Architecture guardrails passed: "
        f"{len(config.line_limits)} line-limit rules, {len(config.forbidden_paths)} forbidden-path rules"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: backend/tools/xtask/src/architecture_guardrails.rs
use std::fs;
use std::path::Path;

use serde::Deserialize;
use thiserror::Error;

pub const DEFAULT_CONFIG_NAME: &str = "architecture_guardrails.toml";

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct LineLimitRule {
    pub path: String,
    pub max_lines: usize,
    pub owner: String,
    pub review_note: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
pub struct ForbiddenPathRule {
    pub path: String,
    pub owner: String,
    pub reason: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Default, Deserialize)]
pub struct GuardrailConfig {
    #[serde(default)]
    pub line_limits: Vec<LineLimitRule>,
    #[serde(default)]
    pub forbidden_paths: Vec<ForbiddenPathRule>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Violation {
    pub code: String,
    pub path: String,
    pub message: String,
}

#[derive(Debug, Error, PartialEq, Eq)]
#[error("{0}")]
pub struct GuardrailConfigError(String);

impl GuardrailConfigError {
    fn new(message: impl Into<String>) -> Self {
        Self(message.into())
    }
}

fn validate_non_empty_text(value: &str, field_name: &str, context: &str) -> Result<(), GuardrailConfigError> {
    if value.trim().is_empty() {
        return Err(GuardrailConfigError::new(format!("{context} is missing required text field {field_name}")));
    }
    Ok(())
}

fn validate_config(config: GuardrailConfig) -> Result<GuardrailConfig, GuardrailConfigError> {
    for (index, rule) in config.line_limits.iter().enumerate() {
        let context = format!("line_limits[{index}]");
        validate_non_empty_text(&rule.path, "path", &context)?;
        if rule.max_lines == 0 {
            return Err(GuardrailConfigError::new(format!("{context} has invalid positive integer field max_lines")));
        }
        validate_non_empty_text(&rule.owner, "owner", &context)?;
        validate_non_empty_text(&rule.review_note, "review_note", &context)?;
    }

    for (index, rule) in config.forbidden_paths.iter().enumerate() {
        let context = format!("forbidden_paths[{index}]");
        validate_non_empty_text(&rule.path, "path", &context)?;
        validate_non_empty_text(&rule.owner, "owner", &context)?;
        validate_non_empty_text(&rule.reason, "reason", &context)?;
    }

    Ok(config)
}

pub fn load_config(config_path: &Path) -> Result<GuardrailConfig, GuardrailConfigError> {
    let raw_text = fs::read_to_string(config_path).map_err(|err| GuardrailConfigError::new(format!("guardrail config not found: {}: {err}", config_path.display())))?;
    let config: GuardrailConfig = toml::from_str(&raw_text).map_err(|err| GuardrailConfigError::new(format!("guardrail config is not valid toml: {}: {err}", config_path.display())))?;
    validate_config(config)
}

fn count_lines(path: &Path) -> Result<usize, std::io::Error> {
    let text = String::from_utf8_lossy(&fs::read(path)?).into_owned();
    Ok(text.lines().count())
}

pub fn evaluate_guardrails(repo_root: &Path, config: &GuardrailConfig) -> Vec<Violation> {
    let mut violations = Vec::new();

    for rule in &config.line_limits {
        let candidate = repo_root.join(&rule.path);
        if !candidate.is_file() {
            violations.push(Violation {
                code: "GUARDED_FILE_MISSING".into(),
                path: rule.path.clone(),
                message: format!("guarded file is missing; either update the architecture guardrail or restore the file (owner: {}; review note: {})", rule.owner, rule.review_note),
            });
            continue;
        }

        let line_count = match count_lines(&candidate) {
            Ok(value) => value,
            Err(err) => {
                violations.push(Violation { code: "GUARDED_FILE_UNREADABLE".into(), path: rule.path.clone(), message: format!("guarded file could not be read: {err}") });
                continue;
            }
        };
        if line_count > rule.max_lines {
            violations.push(Violation {
                code: "MAX_LINES_EXCEEDED".into(),
                path: rule.path.clone(),
                message: format!("{line_count} lines exceeds max {} (owner: {}; review note: {})", rule.max_lines, rule.owner, rule.review_note),
            });
        }
    }

    for rule in &config.forbidden_paths {
        if repo_root.join(&rule.path).exists() {
            violations.push(Violation {
                code: "FORBIDDEN_PATH_PRESENT".into(),
                path: rule.path.clone(),
                message: format!("forbidden path is present (owner: {}; reason: {})", rule.owner, rule.reason),
            });
        }
    }

    violations
}

pub fn format_violations(violations: &[Violation]) -> String {
    let mut lines = vec!["Architecture guardrails failed:".to_string()];
    lines.extend(violations.iter().map(|violation| format!("- [{}] {}: {}", violation.code, violation.path, violation.message)));
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use std::fs;

    use tempfile::tempdir;

    use super::{ForbiddenPathRule, GuardrailConfig, GuardrailConfigError, LineLimitRule, Violation, evaluate_guardrails, format_violations, load_config};

    #[test]
    fn load_config_requires_review_note() {
        let temp_dir = tempdir().expect("tempdir");
        let config_path = temp_dir.path().join("guardrails.toml");
        fs::write(
            &config_path,
            r#"
                [[line_limits]]
                path = "backend/src/example.rs"
                max_lines = 10
                owner = "HeliOS"
            "#,
        )
        .expect("write config");

        let err = load_config(&config_path).expect_err("missing review note should fail");
        assert!(matches!(err, GuardrailConfigError(message) if message.contains("review_note")));
    }

    #[test]
    fn evaluate_reports_oversized_file() {
        let temp_dir = tempdir().expect("tempdir");
        let guarded_file = temp_dir.path().join("backend/src/example.rs");
        fs::create_dir_all(guarded_file.parent().expect("parent")).expect("mkdir");
        fs::write(&guarded_file, "line1\nline2\nline3\n").expect("write guarded file");

        let config = GuardrailConfig {
            line_limits: vec![LineLimitRule { path: "backend/src/example.rs".into(), max_lines: 2, owner: "HeliOS".into(), review_note: "example limit".into() }],
            forbidden_paths: Vec::new(),
        };

        let violations = evaluate_guardrails(temp_dir.path(), &config);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].code, "MAX_LINES_EXCEEDED");
        assert!(violations[0].message.contains("exceeds max 2"));
    }

    #[test]
    fn evaluate_reports_missing_guarded_file() {
        let temp_dir = tempdir().expect("tempdir");
        let config = GuardrailConfig {
            line_limits: vec![LineLimitRule { path: "backend/src/example.rs".into(), max_lines: 10, owner: "HeliOS".into(), review_note: "missing file should fail".into() }],
            forbidden_paths: Vec::new(),
        };

        let violations = evaluate_guardrails(temp_dir.path(), &config);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].code, "GUARDED_FILE_MISSING");
    }

    #[test]
    fn evaluate_reports_forbidden_path() {
        let temp_dir = tempdir().expect("tempdir");
        let forbidden_file = temp_dir.path().join("backend/src/legacy.rs");
        fs::create_dir_all(forbidden_file.parent().expect("parent")).expect("mkdir");
        fs::write(&forbidden_file, "// legacy\n").expect("write forbidden file");

        let config =
            GuardrailConfig { line_limits: Vec::new(), forbidden_paths: vec![ForbiddenPathRule { path: "backend/src/legacy.rs".into(), owner: "HeliOS".into(), reason: "legacy split path".into() }] };

        let violations = evaluate_guardrails(temp_dir.path(), &config);
        assert_eq!(violations.len(), 1);
        assert_eq!(violations[0].code, "FORBIDDEN_PATH_PRESENT");
    }

    #[test]
    fn evaluate_passes_when_repo_matches_rules() {
        let temp_dir = tempdir().expect("tempdir");
        let guarded_file = temp_dir.path().join("frontend/src/example.ts");
        fs::create_dir_all(guarded_file.parent().expect("parent")).expect("mkdir");
        fs::write(&guarded_file, "const x = 1;\n").expect("write guarded file");

        let config = GuardrailConfig {
            line_limits: vec![LineLimitRule { path: "frontend/src/example.ts".into(), max_lines: 5, owner: "HeliOS".into(), review_note: "small demo file".into() }],
            forbidden_paths: vec![ForbiddenPathRule { path: "frontend/src/legacy.ts".into(), owner: "HeliOS".into(), reason: "legacy file must stay deleted".into() }],
        };

        let violations = evaluate_guardrails(temp_dir.path(), &config);
        assert!(violations.is_empty());
    }

    #[test]
    fn format_violations_lists_each_entry() {
        let violations = vec![
            Violation { code: "MAX_LINES_EXCEEDED".into(), path: "a.rs".into(), message: "too big".into() },
            Violation { code: "FORBIDDEN_PATH_PRESENT".into(), path: "b.rs".into(), message: "must stay deleted".into() },
        ];

        let text = format_violations(&violations);
        assert!(text.contains("[MAX_LINES_EXCEEDED] a.rs: too big"));
        assert!(text.contains("[FORBIDDEN_PATH_PRESENT] b.rs: must stay deleted"));
    }
}
